### PR TITLE
Bump LLVM libcxx to 8.0.0

### PR DESCRIPTION
**Pull request by vlaza**
Fixes part **LLVM libcxx 8.0.0** of https://github.com/Microsoft/openenclave/issues/1456

### CODE REVIEW

**Comment by vlaza** [comment]
bors try

**Comment by vlaza** [comment]
bors try

**Comment by CodeMonkeyLeet** [comment]
@vlaza Is this PR complete for libcxx update? Or are you submitting the PR as just the set of 3rdparty changes that builds and intend to submit a separate PR for updating the tests? (I haven't seen any modifications to the libcxx test sets and I see that the Jenkins test run for libcxx tests did fail). 

To clarify the instructions in #1456, I was expecting them to be separate commits in the same PR (see #1498 for reference).

A couple of reasons for marking this as Do Not Merge for now, although you should continue to investigate the test failures:

-  Libcxx usually has a dependency and potential interaction with libc, so we should test this after merging with #1498.

- The release version of LLVM 8.0.0 should be out tomorrow (2/27) so we should wait until that's available and run the comprehensive tests on that.

**Comment by vlaza** [comment]
@CodeMonkeyLeet PR will be extended

**Comment by vlaza** [comment]
It seems that we only got and *rc3* instead of a *final* for http://llvm.org/svn/llvm-project/libcxx/tags/RELEASE_800/
Shall we just proceed with *rc3*? as we are closing to the "code freeze"

**Comment by CodeMonkeyLeet** [comment]
@vlaza I would say hold off on merging to master. If they haven't published a release version, they probably have concerns about quality of the existing release, and we don't have a strong requirement to rush to update libcxx specifically for the march cutoff. 

In the meantime, we should continue to iterate on fixing tests based on the rc3 drop, with the expectation that the delta to the final version should be small once that's ready.

**Comment by vlaza** [comment]
@CodeMonkeyLeet totally agree on that

**Comment by vlaza** [comment]
There still is no *final* release but will test with *rc4* to see changes.

**Comment by vlaza** [comment]
bors try

**Comment by vlaza** [comment]
bors try

**Comment by vlaza** [comment]
800 *final* has been released; I have updated the PR with the changes and testing top-to-bottom.

**Comment by vlaza** [comment]
bors try

**Comment by vlaza** [comment]
@johnkord @CodeMonkeyLeet @ionutbalutoiu @aserdean 
libcxx updated tests lists have 100% PASS rate; do you think we can get this in before branching for release?

File: tests/libcxx/tests.supported [hunk 1]
```diff
@@ -4718,3 +4701,19 @@
...
+../../3rdparty/libcxx/libcxx/test/libcxx/experimental/utilities/ratio/version.pass.cpp
+../../3rdparty/libcxx/libcxx/test/libcxx/experimental/utilities/time/version.pass.cpp
+../../3rdparty/libcxx/libcxx/test/libcxx/experimental/utilities/tuple/version.pass.cpp
+../../3rdparty/libcxx/libcxx/test/libcxx/utilities/utility/pairs/pairs.pair/U_V.pass.cpp
+../../3rdparty/libcxx/libcxx/test/libcxx/utilities/utility/pairs/pairs.pair/const_first_const_second.pass.cpp
+../../3rdparty/libcxx/libcxx/test/libcxx/utilities/utility/pairs/pairs.pair/const_pair_U_V.pass.cpp
+../../3rdparty/libcxx/libcxx/test/libcxx/utilities/utility/pairs/pairs.pair/default.pass.cpp
+../../3rdparty/libcxx/libcxx/test/libcxx/utilities/utility/pairs/pairs.pair/piecewise.pass.cpp
+../../3rdparty/libcxx/libcxx/test/libcxx/utilities/utility/pairs/pairs.pair/rv_pair_U_V.pass.cpp
+../../3rdparty/libcxx/libcxx/test/std/language.support/support.limits/version.pass.cpp
```
Comment (CodeMonkeyLeet) [inline tests/libcxx/tests.supported]: Please sort in alphabetical order in this list

File: tests/libcxx/tests.supported [hunk 2]
```diff
@@ -4718,3 +4701,19 @@
 ../../3rdparty/libcxx/libcxx/test/support/test.support/test_poisoned_hash_helper.pass.cpp
 ../../3rdparty/libcxx/libcxx/test/support/test.workarounds/c1xx_broken_is_trivially_copyable.pass.cpp
 ../../3rdparty/libcxx/libcxx/test/support/test.workarounds/c1xx_broken_za_ctor_check.pass.cpp
+../../3rdparty/libcxx/libcxx/test/libcxx/containers/sequences/vector/vector.cons/construct_iter_iter.pass.cpp
```
Comment (CodeMonkeyLeet) [inline tests/libcxx/tests.supported]: Are these the only new tests that have been added in 8.0.0? Just doing a quick scan, neither of these are accounted for in our test lists:
* test\libcxx\language.support\support.dynamic\libcpp_deallocate.sh.cpp
* test\std\containers\associative\multiset\insert_emplace_allocator_requirements.pass.cpp

I suspect there are probably others? Again, I would do a full recursive ls of the new test folder, sort the results and diff them against our tests.all to get a comprehensive view of everything that's changed.

Also, any new tests should be added to the tests.all list in addition to either tests.support/tests.unsupported/tests.broken.

File: tests/libcxx/tests.supported (hunk 1 above)
Comment (vlaza) [reply tests/libcxx/tests.supported]: Tests were updated and sorted.

File: tests/libcxx/tests.supported (hunk 2 above)
Comment (vlaza) [reply tests/libcxx/tests.supported]: Rebuilt test lists with full recursive scan and reiterated until build was clean have a stable list of passing tests.

**Comment by vlaza** [comment]
bors try

File: tests/libcxx/tests.all [hunk 3]
```diff
@@ -5375,19 +6221,30 @@
...
 ../../3rdparty/libcxx/libcxx/test/support/test.support/test_convertible_header.pass.cpp
 ../../3rdparty/libcxx/libcxx/test/support/test.support/test_demangle.pass.cpp
+../../3rdparty/libcxx/libcxx/test/support/test.support/test_macros_header_exceptions.fail.cpp
 ../../3rdparty/libcxx/libcxx/test/support/test.support/test_macros_header_exceptions.pass.cpp
+../../3rdparty/libcxx/libcxx/test/support/test.support/test_macros_header_rtti.fail.cpp
 ../../3rdparty/libcxx/libcxx/test/support/test.support/test_macros_header_rtti.pass.cpp
 ../../3rdparty/libcxx/libcxx/test/support/test.support/test_poisoned_hash_helper.pass.cpp
 ../../3rdparty/libcxx/libcxx/test/support/test.workarounds/c1xx_broken_is_trivially_copyable.pass.cpp
 ../../3rdparty/libcxx/libcxx/test/support/test.workarounds/c1xx_broken_za_ctor_check.pass.cpp
+../../3rdparty/libcxx/libcxx/utils/google-benchmark/cmake/gnu_posix_regex.cpp
```
Comment (CodeMonkeyLeet) [inline tests/libcxx/tests.all]: This last block also don't look like tests, we should drop these.

File: tests/libcxx/tests.all [hunk 4]
```diff
@@ -1,11 +1,67 @@
+../../3rdparty/libcxx/libcxx/benchmarks/algorithms.bench.cpp
```
[the remark this answers is not in the document]
Comment (vlaza) [reply tests/libcxx/tests.all]: L1-L51 inclusive? or the block until "../../3rdparty/libcxx/libcxx/test/libcxx/algorithms/alg.modifying.operations/alg.random.shuffle/random_shuffle.cxx1z.pass.cpp" ?

File: tests/libcxx/tests.all (hunk 3 above)
Comment (vlaza) [reply tests/libcxx/tests.all]: Shall I remove the whole block ?
../../3rdparty/libcxx/libcxx/utils/google-benchmark/cmake/gnu_posix_regex.cpp
../../3rdparty/libcxx/libcxx/utils/google-benchmark/cmake/posix_regex.cpp
../../3rdparty/libcxx/libcxx/utils/google-benchmark/cmake/std_regex.cpp
../../3rdparty/libcxx/libcxx/utils/google-benchmark/cmake/steady_clock.cpp
../../3rdparty/libcxx/libcxx/utils/google-benchmark/cmake/thread_safety_attributes.cpp

Comment (CodeMonkeyLeet) [reply tests/libcxx/tests.all]: Yes … these aren't actually tests and there's not point having them in tests.all and tests.unsupported.

Comment (vlaza) [reply tests/libcxx/tests.all]: Removed from tests.all , tests.unsupported and tests.supported (there was one google-benchmark there also)

File: tests/libcxx/tests.supported [hunk 5]
```diff
@@ -1,10 +1,15 @@
+../../3rdparty/libcxx/libcxx/fuzzing/fuzz_test.cpp
```
[the remark this answers is not in the document]
Comment (vlaza) [reply tests/libcxx/tests.supported]: Removed from tests.supported and tests.all

File: tests/libcxx/tests.all (hunk 4 above)
[the remark this answers is not in the document]
Comment (vlaza) [reply tests/libcxx/tests.all]: Removed from tests.all L1-L51 as they did not begin with "../../3rdparty/libcxx/libcxx/test/"

**Comment by vlaza** [comment]
bors try

**Comment by vlaza** [comment]
@CodeMonkeyLeet @johnkord @ionutbalutoiu 
libcxx tests passed

**Comment by CodeMonkeyLeet** [comment]
@vlaza I sent you a spreadsheet that where I’m still finding discrepancies between the groups of tests. Some of these look like they may have been carried forward from previous updates, but can you make sure that they are all consistent?

- 445 tests exist in `tests.all` but have not been assigned to `tests.supported`, `tests.unsupported` or `tests.broken`.
- 56 tests exist in `tests.unsupported` but are not declared in `tests.all`. It may be that these tests no longer exist in the latest LLVM version and should be removed, can you please verify?

Also, for any tests that have regressed (moved from tests.supported to tests.broken) like `test/libcxx/algorithms/half_positive.pass.cpp`, can you make sure that we have an issue to track it after the PR is merged?

**Comment by vlaza** [comment]
@CodeMonkeyLeet I have not received any file from you on any of my email addresses but will look into.
tests.all has been rebuilt from scratch by recursive scan tests in 800 release.

_445 tests exist in tests.all but have not been assigned to tests.supported, tests.unsupported or tests.broken._ will do cross check.

_56 tests exist in tests.unsupported but are not declared in tests.all. It may be that these tests no longer exist in the latest LLVM version and should be removed, can you please verify?_ will double check (for sure they must be removed from tests.unsupported also.

**Comment by vlaza** [comment]
@CodeMonkeyLeet I received the spreadsheet via @ionutbalutoiu and took it into account for tests.* redistribution.
tests.all = 6194
tests.supported = 4882
tests.unsupported = 1311
tests.broken = 1

**Comment by vlaza** [comment]
bors try

**Comment by CodeMonkeyLeet** [comment]
@vlaza Thanks for reconciling the various test sets. I'm currently trying to evaluate the new set of tests moved into `tests.unsupported` and seeing if there's a way we can rationalize that.

@anakrish FYI, you might want to take a look at the broken `test/libcxx/algorithms/half_positive.pass.cpp` which is newly failing to compile in debug.

**Comment by vlaza** [comment]
bors try

**Comment by vlaza** [comment]
@CodeMonkeyLeet @johnkord @ionutbalutoiu I have made small adjustments and re-based this PR; did also re-run tests in this formula multiple times and we are green.

**Comment by anitagov** [comment]
This is great news. Do all the 5000+ tests PASS with the ENABLE_FULL_LIBCXX_TESTS flag?

**Comment by vlaza** [comment]
@anitagov Yes, all the 5000+ tests Pass.

**Comment by andschwa** [comment]
Blocked until IO changes go in. Thanks!

**Comment by vlaza** [comment]
Thank you @aserdean for breaking into smaller commits.

**Comment by vlaza** [comment]
We have re-tested the PR and looks good.

**Comment by andschwa** [comment]
Did the 8.0.0 release happen?

**Comment by anitagov** [comment]
@andschwa - LLVM 8.0.0 was released on March 20, 2019.

**Comment by ionutbalutoiu** [comment]
@andschwa @johnkord @CodeMonkeyLeet

Are all the I/O changes merged ? I noticed that https://github.com/microsoft/openenclave/pull/1772 merged.
If there isn't anything else blocking this, can we merge it ?

It's all good to go from our side.

**Comment by anitagov** [comment]
This should have gone into Release 0.6. Definitely missed out.
@yakman2020 @CodeMonkeyLeet - Can we get this into Release 0.6?

**Comment by CodeMonkeyLeet** [comment]
@anitagov There's no requirement for it to be in v0.6 AFAIK. It can go into master for v0.7 at this point.

**Comment by ionutbalutoiu** [comment]
bors try

**Comment by ionutbalutoiu** [comment]
bors try

**Comment by ionutbalutoiu** [comment]
I double checked with Bors, and the PR still looks in good shape to be merged (the initial `try` had a transient error when validating SSL certificates, but it didn't reproduce on the second run).

If the merges into `master` for the `v0.7` branch are unblocked, I suggest merging this.

/cc @johnkord @anitagov @CodeMonkeyLeet @andschwa

**Comment by CodeMonkeyLeet** [comment]
Superseded by #2026

**Pull request closed without merging**